### PR TITLE
BugFix in `SelectSwapQROM` when multiple datasets are specified

### PR DESCRIPTION
**Pull request by tanujkhattar**
Two bugfixes:

1. Use `self.use_dirty_ancilla` in `_value_equality_values_` -- This is important to make sure `bloq.t_complexity()` returns different values when `bloq` is initialized with `use_dirty_ancilla=True` vs `use_dirty_ancilla=False`. Right now, they can return the same value because the t-complexity is cached
2. `qrom_batched_target[0]` can return a numpy array because the target shapes can be multi-dimensional. This is especially true when input data is multi-dimensional (i.e. `data[N][M] --> bached_qrom_data[N / k1][M / k2]`). We need the `target_0_0_..._0`'th register and not the `target_0`th register; which can be multidimensional. 

The added tests fail on main but pass with this PR.

### CODE REVIEW

File: qualtran/bloqs/data_loading/select_swap_qrom_test.py [hunk 1]
```diff
@@ -158,3 +164,17 @@ def test_qroam_many_registers():
         log_block_sizes=(log_block_size,),
     )
     qrom.call_graph()
+
+
+@pytest.mark.parametrize('bloq_example', [_qroam_multi_data, _qroam_multi_dim])
```
Comment (fdmalone) [inline qualtran/bloqs/data_loading/select_swap_qrom_test.py:169]: Why isn't this just a regular old bloq_example auto tester? Won't it run more tests?

**Review by fdmalone — APPROVED** [review]

**Comment by fdmalone** [comment]
> We need the `target_0_0_..._0`'th register and not the `target_0`th register; which can be multidimensional. 

I was a little confused by the underscores here given our weird target register convention, you just mean you want the actual zero'th target register not the zero'th row (in the 2D example)

File: qualtran/bloqs/data_loading/select_swap_qrom_test.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/data_loading/select_swap_qrom_test.py]: Changed to `auto tester`. Found and fixed another bug in `nth_operation_callgraph` of `QROM`; which was not updated to support the recently added `target_shapes` feature required for `SelectSwapQROM`.

**Comment by tanujkhattar** [comment]
> I was a little confused by the underscores here given our weird target register convention, you just mean you want the actual zero'th target register not the zero'th row (in the 2D example)

Yup

**Comment by mpharrigan** [comment]
Why are we using value_equality_values at all?

**Comment by tanujkhattar** [comment]
> Why are we using value_equality_values at all?

We store tuples of numpy arrays, which don't support value equality by default.

**Comment by tanujkhattar** [comment]
@fdmalone @mpharrigan Removed the use of `_value_equality_values_` with the fancy attrs functionality Fionn mentioned above. This is ready to go now.

**Pull request merged**